rfcs/20180920-unify-rnn-interface.md

```diff
@@ -0,0 +1,377 @@
+       
+### RNN Cell
+Following table lists out the current status for RNN cells between 2 APIs:
+
+| Keras RNN | TF RNN (tf.nn.rnn_cell) | Comment |
+|:----------|:----------------------- |:------  |
+| |RNNCell|Abstract base class.|
+| |LayerRNNCell|Not publicly exposed. Layer support for RNNCell.|
+|SimpleRNNCell|BasicRNNCell|Identical implementation of vanilla RNN, the underlying weights are structured differently.|
+|LSTMCell|BasicLSTMCell|No peephole, clipping, projection. Keras allows kernel_activation to be customized (default=hard_sigmoid)|
+| |LSTMCell|Support peephole, clipping and projection.|
+|GRUCell|GRUCell|Identical implementation, the underlying weights are structured differently.|
+|StackedRNNCells|MultiRNNCell|Identical implementation|
+| |DeviceWrapper| |
+| |DropoutWrapper|Keras support dropout at the cell level and configured by the init parameters. All of the Keras cell support input and state dropout. DropoutWrapper support output dropout as well. Both Keras and DropoutWrapper support variational dropout|
+| |ResidualWrapper| |
+|BidirectionalWrapper| | |
+
+Apart from the cells in tf.nn.rnn_cell, there are more cells/wrappers in tf.contrib.rnn. Part of the 
+cell/wrapper are duplicate with existing tf.nn.rnn_cell, and the rest part provides extra 
+functionality. See sections below.
+
+The interface between TF RNN and Keras RNN APIs are actually similar.
+
+Both of them requires:
+
+```python
+def call(self, inputs, prev_state):
+  # Calculate the output and new_state based on the cell type.
+  return output, new_state
+```
+
+```python
+@property	
+def state_size(self):
+  # The size of the internal state for each step. 
+  return state_size
+```
+
+In addition, TF RNN cell requires:
+
+```python
+@property
+def output_size(self):
+  # The size of output.
+
+
+def zero_state(self, batch_size, dtype):
+   # return the initial state for timestep 0, which is used when caller didn't 
+   # specify any initial state when constructing the layer.
+```
+
+Keras RNN cell requires cell to implement the following method:
+
+```python
+def get_config(self):
+  # return the dict of the cell attribute, which can be used to rebuild the cell.
+```
+
+#### Changes to make
+1. Update the Keras RNN cell to support the extra two methods required for TF RNN cell. This will 
+enable all the cells to be used by both layer interfaces.
+1. Deprecate the duplicated RNN cells in TF RNN and suggest user to use the Keras equivalent, 
+namely:
+    1. nn.BasicRNNCell -> keras.SimpleRNNCell
+    1. (nn.BasicLSTMCell,  keras.LSTMCell) -> nn.LSTMCell
+    1. nn.GRUCell -> keras.GRUCell
+    1. nn.MultiRNNCell -> keras.StackedRNNCells
+    1. (Optional) Remove the individual dropout param from Keras cells and preferred the DropWrapper 
+        from TF RNN.
+
+1. All the RNN cells need to have a unified namespace. 
+
+The side effects of deleting cell is that previously saved checkpoint will not work, and maybe TF 
```

Comment:
I don't follow, which `cell` is being removed?

Perhaps it should read `deleting the dropout parameter` instead of `deleting cell`?

Reply:
As listed above, eg tf.nn.BasicRNNCell will be removed and tf.keras.SimpleRNNCell will be preferred cell that achieve the same behavior.

Reply:
Thank you for the clarification. I had misunderstood the point; I thought that the API would be deprecated _when_ 2.0 is released.